crates/bevy_ecs/src/query/iter.rs

```diff
@@ -69,73 +47,156 @@ where
-                        self.current_index += 1;
-                        continue;
-                    }
+        // MaybeUninit::uninit_array is unstable
+        let mut cursors: [MaybeUninit<QueryIterationCursor<'s, Q, F>>; K] =
+            MaybeUninit::uninit().assume_init();
+        for i in 0..K {
+            match i {
+                0 => cursors[i].as_mut_ptr().write(QueryIterationCursor::init(
+                    world,
+                    query_state,
+                    last_change_tick,
+                    change_tick,
+                )),
+                _ => cursors[i]
+                    .as_mut_ptr()
+                    .write(QueryIterationCursor::init_empty(
+                        world,
+                        query_state,
+                        last_change_tick,
+                        change_tick,
+                    )),
+            }
+        }
 
-                    let item = self.fetch.archetype_fetch(self.current_index);
-                    self.current_index += 1;
-                    return Some(item);
+        // MaybeUninit::array_assume_init is unstable
+        let cursors: [QueryIterationCursor<'s, Q, F>; K] =
+            (&cursors as *const _ as *const [QueryIterationCursor<'s, Q, F>; K]).read();
+
+        QueryPermutationIter {
+            world,
+            query_state,
+            tables: &world.storages().tables,
+            archetypes: &world.archetypes,
+            cursors,
+        }
+    }
+}
+
+impl<'w, 's, Q: WorldQuery, F: WorldQuery, const K: usize> Iterator
+    for QueryPermutationIter<'w, 's, Q, F, K>
+where
+    F::Fetch: FilterFetch,
+    Q::Fetch: Clone,
+    F::Fetch: Clone,
+{
+    type Item = [<Q::Fetch as Fetch<'w>>::Item; K];
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        unsafe {
+            // first, iterate from last to first until next item is found
+            'outer: for i in (0..K).rev() {
+                match self.cursors[i].next(&self.tables, &self.archetypes, &self.query_state) {
+                    Some(_) => {
+                        // walk forward up to last element, propagating cursor state forward
+                        for j in (i + 1)..K {
+                            self.cursors[j] = self.cursors[j - 1].clone();
+                            match self.cursors[j].next(
+                                &self.tables,
+                                &self.archetypes,
+                                &self.query_state,
+                            ) {
+                                Some(_) => {}
+                                None if i > 0 => continue 'outer,
+                                None => return None,
+                            }
+                        }
+                        break;
+                    }
+                    None if i > 0 => continue,
+                    None => return None,
                 }
```

Comment:
I'm not a fan of this complex double loop, and I believe this can be written simpler. Just wasn't able to simplify it more for now. If you have any idea how to do that, I'm all ears.